pkg/api/v1/marshal.go

```diff
@@ -0,0 +1,38 @@
+package v1
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+func (ip *IdentityProvider) UnmarshalJSON(b []byte) error {
+	dummy := struct {
+		Name     string          `json:"name,omitempty"`
+		Provider json.RawMessage `json:"provider,omityempty"`
+	}{}
+	err := json.Unmarshal(b, &dummy)
+	if err != nil {
+		return err
+	}
+	// peek inside to find out type
+	m := map[string]interface{}{}
+	err = json.Unmarshal(dummy.Provider, &m)
+	if err != nil {
+		return err
+	}
+
+	switch m["kind"] {
+	case "AADIdentityProvider":
+		ip.Provider = &AADIdentityProvider{}
+		//unmarshal to the right type
+		err = json.Unmarshal(dummy.Provider, &ip.Provider)
```

Comment:
No need to use a reference if ip.Provider is a pointer already?

Reply:
True, but I think `json.Unmarshal(foo, &bar)` conforms to the principle of least surprise - IMO it always behaves as expected, whereas `json.Unmarshal(foo, bar)` you have to check extra carefully.

Reply:
TIL